extensions/mypy_extensions.py

```diff
@@ -7,16 +7,76 @@
 
-def TypedDict(typename, fields):
-    """TypedDict creates a dictionary type that expects all of its
+
+def _dict_new(inst, cls, *args, **kwargs):
+    return dict(*args, **kwargs)
+
+def _check_fails(cls, other):
+    try:
+        if sys._getframe(1).f_globals['__name__'] not in ['abc', 'functools']:
+            raise TypeError('TypedDict does not support instance and class checks')
+    except (AttributeError, ValueError):
+        pass
+    return False
+
+class _TypedDictMeta(type):
+    def __new__(cls, name, bases, ns):
+        tp_dict = super(_TypedDictMeta, cls).__new__(cls, name, (dict,), ns)
+        try:
+            tp_dict.__module__ = sys._getframe(2).f_globals.get('__name__', '__main__')
+        except (AttributeError, ValueError):
+            pass
+        anns = ns.get('__annotations__', {})
+        msg = "TypedDict('Name', {f0: t0, f1: t1, ...}); each t must be a type"
+        anns = {n: _type_check(tp, msg) for n, tp in anns.items()}
+        for base in bases:
+            anns.update(base.__dict__.get('__annotations__', {}))
+        tp_dict.__annotations__ = anns
+        if tp_dict.__name__ != 'TypedDict':
+            tp_dict.__new__ = types.MethodType(_dict_new, tp_dict)
+        return tp_dict
+
+    __instancecheck__ = __subclasscheck__ = _check_fails
+
+
+class _TypedDict(object):
+    """A simple typed name space. At runtime it is equivalent to a plain dict.
+
+    TypedDict creates a dictionary type that expects all of its
     instances to have a certain set of keys, with each key
     associated with a value of a consistent type. This expectation
     is not checked at runtime but is only enforced by typecheckers.
+    Usage::
+
+        Point2D = TypedDict('Point2D', {'x': int, 'y': int, 'label': str})
+        a: Point2D = {'x': 1, 'y': 2, 'label': 'good'}  # OK
+        b: Point2D = {'z': 3, 'label': 'bad'}           # Fails type check
+        assert Point2D(x=1, y=2, label='first') == dict(x=1, y=2, label='first')
+
+    The type info could be accessed via Point2D.__annotations__. TypedDict
+    supports two additional equivalent forms::
+
+        Point2D = TypedDict('Point2D', x=int, y=int, label=str)
+
+        class Point2D(TypedDict):
+            x: int
+            y: int
+            label: str
+
+    The latter syntax is only supported in Python 3.6+
     """
-    def new_dict(*args, **kwargs):
-        return dict(*args, **kwargs)
+    def __new__(cls, _typename, _fields=None, **kwargs):
+        if _fields is None:
+            _fields = kwargs
+        elif kwargs:
+            raise TypeError("Either list of fields or keywords"
+                            " can be provided to TypedDict, not both")
+        return cls.__class__(_typename, (), {'__annotations__': dict(_fields)})
+
 
-    new_dict.__name__ = typename
-    new_dict.__supertype__ = dict
-    return new_dict
+TypedDict = _TypedDictMeta('TypedDict', _TypedDict.__bases__, dict(_TypedDict.__dict__))
```

Comment:
I'm guessing this awkward way of defining the class is to get around the different metaclass syntaxes for Python 2 and 3. But it makes my head explode trying to understand what's happening. I *think* that one of the problems is actually that `TypedDict` is itself a metaclass. (We see this when we realize that *instances* of TypedDict are really *classes* whose instances are dicts.) So maybe `_TypedDict` needs to be combined into `TypedDictMeta`? What tipped me off was that `_TypedDict.__new__` ends up not calling `object()` or `super().__new__()` but `cls.__class__()`, which is really a fancy way of spelling `TypedDictMeta()`.